src/chains.rs

```diff
@@ -92,12 +92,13 @@ fn format_chain_item(
     context: &RewriteContext<'_>,
     rewrite_shape: Shape,
     allow_overflow: bool,
-) -> Option<String> {
+) -> RewriteResult {
+    // TODO this should only match RewriteError::ExceedsMaxWidth
     if allow_overflow {
-        item.rewrite(context, rewrite_shape)
-            .or_else(|| format_overflow_style(item.span, context))
+        item.rewrite_result(context, rewrite_shape)
+            .or_else(|_| format_overflow_style(item.span, context).unknown_error())
     } else {
-        item.rewrite(context, rewrite_shape)
+        item.rewrite_result(context, rewrite_shape)
     }
```

Comment:
My initial thought was that we might want to retry formatting by applying `format_overflow_style` only when `item.rewrite_result()` returns `RewriteError::ExceedsMaxWidth`. Since it'll retry formatting only when `allow_overflow` is true, I thought calling `format_overflow_style` on another rw error besides max width might be inappropriate. 

Though, considering that (1) we should not distinguish the type of the err to preserve original formatting mechanism and (2) some of unknown errors might indicate max width error (ex. `Expr` does not have its rw_result implementation yet),
I left the code as is.  

If we'll gonna provide the original input from given span no matter what kind of err occurs during `item.rewrite_result`, I think I can remove this comment.

Reply:
I think for now we can keep the current implementation of the code as is, though I think the idea of only calling `format_overflow_style` when we're dealing with a `RewriteError::ExceedsMaxWidth` error is an improvement we can make in the future. If that's what the comment above is referring to then I'd suggest capturing those details there. Also, it might be clearer if the comment was placed closer to the `format_overflow_style` call.